sycl/include/sycl/handler.hpp

```diff
@@ -791,23 +791,32 @@ class __SYCL_EXPORT handler {
         "-fsycl-host-compiler-options='/std:c++latest' "
         "might also help.");
 #endif
-    // Empty name indicates that the compilation happens without integration
-    // header, so don't perform things that require it.
-    if constexpr (KernelHasName) {
-      // TODO support ESIMD in no-integration-header case too.
-
-      clearArgs();
-      extractArgsAndReqsFromLambda(MHostKernel->getPtr(),
-                                   &(detail::getKernelParamDesc<KernelName>),
-                                   detail::getKernelNumParams<KernelName>(),
-                                   detail::isKernelESIMD<KernelName>());
-      MKernelName = detail::getKernelName<KernelName>();
+    // If a kernel has no special captures, we can extract the arguments
+    // from the kernel object directly when the kernel is enqueued.
+    if constexpr (detail::hasSpecialCaptures<KernelName>()) {
+      // Empty name indicates that the compilation happens without integration
+      // header, so don't perform things that require it.
+      if constexpr (KernelHasName) {
+        // TODO support ESIMD in no-integration-header case too.
+
+        clearArgs();
+        extractArgsAndReqsFromLambda(MHostKernel->getPtr(),
+                                     &(detail::getKernelParamDesc<KernelName>),
+                                     detail::getKernelNumParams<KernelName>(),
+                                     detail::isKernelESIMD<KernelName>());
+        MKernelName = detail::getKernelName<KernelName>();
+      } else {
+        // In case w/o the integration header it is necessary to process
+        // accessors from the list(which are associated with this handler) as
+        // arguments. We must copy the associated accessors as they are checked
+        // later during finalize.
+        setArgsToAssociatedAccessors();
+      }
     } else {
-      // In case w/o the integration header it is necessary to process
-      // accessors from the list(which are associated with this handler) as
-      // arguments. We must copy the associated accessors as they are checked
-      // later during finalize.
-      setArgsToAssociatedAccessors();
+      MKernelName = detail::getKernelName<KernelName>();
+      prepareForDirectArgumentCopy((const void *)&KernelFunc,
+                                   detail::getKernelNumParams<KernelName>(),
+                                   &detail::getKernelParamDesc<KernelName>);
```

Comment:
This is wrong in its current version, I think. Line 768 might have left `KernelFunc` in moved-out state. How exactly is this beneficial? We've already created a copy in `MHostKernel`, what extra overhead are you trying to eliminate?